src/drivers/magnetometer/qmc5883p/QMC5883P.cpp

```diff
@@ -0,0 +1,308 @@
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "QMC5883P.hpp"
+
+using namespace time_literals;
+
+static constexpr int16_t combine(uint8_t msb, uint8_t lsb)
+{
+	return (msb << 8u) | lsb;
+}
+
+QMC5883P::QMC5883P(const I2CSPIDriverConfig &config) :
+    I2C(config),
+    I2CSPIDriver(config),
+    _px4_mag(get_device_id(), config.rotation)
+{
+}
+
+QMC5883P::~QMC5883P()
+{
+    perf_free(_reset_perf);
+    perf_free(_bad_register_perf);
+    perf_free(_bad_transfer_perf);
+}
+
+int QMC5883P::init()
+{
+    int ret = I2C::init();
+
+    if (ret != PX4_OK){
+        DEVICE_DEBUG("I2C::init failed (%i)", ret);
+        return ret;
+    }
+
+    return Reset() ? 0 : -1;
+}
+
+bool QMC5883P::Reset()
+{
+    _state = STATE::RESET;
+    ScheduleClear();
+    ScheduleNow();
+    return true;
+}
+
+void QMC5883P::print_status()
+{
+	I2CSPIDriverBase::print_status();
+
+	perf_print_counter(_reset_perf);
+	perf_print_counter(_bad_register_perf);
+	perf_print_counter(_bad_transfer_perf);
+}
+
+int QMC5883P::probe()
+{
+	_retries = 1;
+
+	for (int i = 0; i < 3; i++) {
+		// first read 0x0 once
+		const uint8_t cmd = 0;
+		uint8_t buffer{};
+
+		if (transfer(&cmd, 1, &buffer, 1) == PX4_OK) {
+			const uint8_t CHIP_ID = RegisterRead(Register::CHIP_ID);
+
+			if (CHIP_ID == Chip_ID) {
+				return PX4_OK;
+			}
+		}
+	}
+
+	return PX4_ERROR;
+}
+
+void QMC5883P::RunImpl()
+{
+    const hrt_abstime now = hrt_absolute_time();
+
+    switch (_state) {
+    case STATE::RESET:
+        {
+	// CNTL2: Software Reset
+        RegisterWrite(Register::CNTL2, CNTL2_BIT::SOFT_RST);
+	_reset_timestamp = now;
+        _failure_count = 0;
+        _state = STATE::WAIT_FOR_RESET;
+        perf_count(_reset_perf);
+        ScheduleDelayed(100_ms); // POR Completion Time
+        break;
+	}
+    case STATE::WAIT_FOR_RESET:
+	{
+        	// SOFT_RST: This bit is automatically reset to zero after POR routine
+		if ((RegisterRead(Register::CHIP_ID) == Chip_ID)
+		    && ((RegisterRead(Register::CNTL2) & CNTL2_BIT::SOFT_RST) == 0)) {
+
+			// if reset succeeded then configure
+			_state = STATE::CONFIGURE;
+			ScheduleDelayed(10_ms);
+
+		} else {
+			// RESET not complete
+			if (hrt_elapsed_time(&_reset_timestamp) > 1000_ms) {
+				PX4_DEBUG("Reset failed, retrying");
+				_state = STATE::RESET;
+				ScheduleDelayed(100_ms);
+
+			} else {
+				PX4_DEBUG("Reset not complete, check again in 10 ms");
+				ScheduleDelayed(10_ms);
+			}
+		}
+
+		break;
+	}
+    case STATE::CONFIGURE:
+        if (Configure()) {
+            // if configure succeeded then start reading every 20 ms (50 Hz)
+            _state = STATE::READ;
+            ScheduleOnInterval(20_ms, 20_ms);
+
+        } else {
+                // CONFIGURE not complete
+			    if (hrt_elapsed_time(&_reset_timestamp) > 1000_ms) {
```

Comment:
```suggestion
			    if (hrt_elapsed_time(&_reset_timestamp) > 1_s) {
```
same comment as above about the time delay. Can you check throughout?

Reply:
> same comment as above about the time delay. Can you check throughout?

In the latest submission, I have completed the revisions. Please review it.